[AME-842] Support product information on the template

elmiranda opened the pull request:
**DO NOT MERGE BEFORE GA**

https://reddit.atlassian.net/browse/AME-842

Adds support for products as a simple table in a default-collapsed "Additional Information" section. 

---
~~I need to confirm what the copy should look like, but I don't think the layout (logic) will change.~~
There are some logical updates in https://github.com/reddit/reddit-gtm-template/commit/82df088a4f7fce8255a8ac17b661b01dbfa50fd7.
Checked with Katie and Anshula and the copy should be A-OK 👍

---

NOTE: Changes in https://github.com/reddit/reddit-gtm-template/pull/17/commits/82df088a4f7fce8255a8ac17b661b01dbfa50fd7 and https://github.com/reddit/reddit-gtm-template/pull/17/commits/9d6d72f9cdea4eb15d83bbc1a057d1e3eee13cf4 require updates to the pixel to be deployed for usage. ✅

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,43 @@ ___TEMPLATE_PARAMETERS___
+            "valueHint": ""
+          }
+        ],
+        "help": "Configure product-level information with Reddit. Use one row per unique product. The Product ID and Product Category are required, but the Product Name is optional."
```
sn00bie question: the category is required?  I would've figured it'd be optional, and that the only required field is the ID.

elmiranda replied:
Yeah a category is required because we don't want to ingest any information that is potentially sensitive. If we put forth that we expect the advertisers to be truthful, by filtering out anything with a sensitive category, we absolve ourselves of liability of ingesting sensitive information.

For Alpha, it's technically not required because we'll be working with non-sensitive advertisers, but I'd rather just do it once than special-case it for Alpha.

woodchuckgym replied:
Ahhhh, I totally missed that this is just the help text.  👍   Nevermind me.

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,43 @@ ___TEMPLATE_PARAMETERS___
+            "displayName": "Product Category",
+            "name": "category",
+            "type": "TEXT",
+            "valueHint": ""
```
Do we wanna provide a hint as to what category names are acceptable?  While ultimately this'll be free form -- right? -- for the first version of shopping ads are we using the IAB content taxonomy?

elmiranda replied:
Yeah good call!

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,43 @@ ___TEMPLATE_PARAMETERS___
+            "displayName": "Product ID",
+            "name": "id",
+            "type": "TEXT",
+            "valueHint": ""
```
Do we wanna provide a hint that this ID should match the item's ID in the catalog?

elmiranda replied:
Good idea!

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,43 @@ ___TEMPLATE_PARAMETERS___
     ]
+  },
+  {
+    "type": "GROUP",
+    "name": "additionalInformation",
+    "groupStyle": "ZIPPY_CLOSED",
```
The section title here is "Additional Information", and then there's a nested table for "Product Information".  Will there ever be any other sections within the "Additional Information"?  If not, why not simply make the "Product Information" the whole section?

elmiranda replied:
It's a little bit of a chicken and egg situation. I wanted the outer `Additional Information` group so that it could be collapsed by default. But that section isn't properly supporting Help Text, so I added the `Product Information` name to the nested table so that it would display the help text. But I didn't want to have it show `Product Information` twice...

woodchuckgym commented on template.tpl:
```diff
@@ -365,17 +402,28 @@ if (!data.enableFirstPartyCookies) {
-  currency: data.currency,
-  value: data.transactionValue,
-};
+var eventMetadata = {};
+
+if (data.currency) {
+  eventMetadata.currency = data.currency;
+}
+
+if (data.transactionValue) {
+  eventMetadata.value = data.transactionValue;
+}
```
```suggestion
var eventMetadata = {
  currency: data.currency,
  value: data.transactionValue,
};
```
Falsy values will already get excluded by the pixel.js.  There's no need for extra tests here.

woodchuckgym replied:
Oh, and there's a bug in the second test.  A zero value might be meaningful but is falsy.  We don't want to exclude it.

elmiranda replied:
Ahh true about the falsy value. OK I will change this back.

woodchuckgym commented on template.tpl:
```diff
@@ -365,17 +402,28 @@ if (!data.enableFirstPartyCookies) {
 }
 
-if (data.eventType != "SignUp" && data.eventType != "Lead") {
+if (data.eventType != "SignUp" && data.eventType != "Lead" && data.itemCount) {
```
This is a bug.  A zero value is falsy but is meaningful.  We don't want to exclude it.

woodchuckgym commented on template.tpl:
```diff
@@ -365,17 +402,28 @@ if (!data.enableFirstPartyCookies) {
 
 // Certain events don't support certain params, so we conditionally set them
-if (data.eventType != "AddToCart" && data.eventType != "AddToWishlist") {
+if (data.eventType != "AddToCart" && data.eventType != "AddToWishlist" && data.transactionId) {
```
Falsy values such as `null` or `""` will be excluded by the pixel.js.  There's no need for this check here.

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,44 @@ ___TEMPLATE_PARAMETERS___
+            "displayName": "Product Category",
+            "name": "category",
+            "type": "TEXT",
+            "valueHint": "e.g. IAB content taxonomy",
```
```suggestion
            "valueHint": "e.g., IAB content taxonomy",
```
Slight grammar typo.

My preference would be for writing out the words to make it as broadly understandable as possible, but YMMV.

woodchuckgym commented on template.tpl:
```diff
@@ -318,6 +318,44 @@ ___TEMPLATE_PARAMETERS___
+            "name": "category",
+            "type": "TEXT",
+            "valueHint": "e.g. IAB content taxonomy",
+            "isUnique": false
```
Nice!  And another sn00bie question: should we make the product name non-unique as well?  I s'pose I'm thinking that we should have the ID be unique for sure.  But what if one product ID is for a blue tuna can, and another product ID is for a red tuna can, but the product name for both is "tuna can"?

elmiranda replied:
Actually, this line isn't supposed to be here at all. I think I accidentally had the advanced settings open when I saved the template. If it's not preset, there's no uniqueness constraint 👍

elmiranda commented:
Re-requesting review as there are some logical updates in https://github.com/reddit/reddit-gtm-template/pull/17/commits/82df088a4f7fce8255a8ac17b661b01dbfa50fd7. Checked with Katie and Anshula and the copy should be A-OK 👍

woodchuckgym commented on template.tpl:
```diff
@@ -367,6 +408,7 @@ var setInWindow = require('setInWindow');
 var callInWindow = require('callInWindow');
 var createQueue = require('createQueue');
 var makeTableMap = require('makeTableMap');
+var JSON = require('JSON');
```
Ah, no wonder there's a `require` for the native object: `___SANDBOXED_JS_FOR_WEB_TEMPLATE___`.

woodchuckgym commented on template.tpl:
```diff
@@ -407,10 +449,16 @@ var eventMetadata = {
+} else if (data.productInputType == "entryJSON" && data.productsJSON && data.productsJSON.length) {
+  // The text input should represent JSON. We need to parse it.
+  // If it fails, products simply won't be defined.
+  var products = JSON.parse(data.productsJSON);
```
```
var products = [];
try {
  products = JSON.parse(data.productsJSON);
} catch (e) {
  console.error("Malformed products data.  Skipping ...");
}
```
Otherwise an exception will get thrown with malformed data.  The script will stop processing, and the pixel won't fire.

try/catch should be broadly compatible with browsers: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#browser_compatibility

woodchuckgym reviewed (changes requested):
Heya.  I think the `JSON.parse()` should be wrapped in a try/catch.

More generally, do you have a test page set up with the pixel to test this with?

elmiranda commented:
> Heya. I think the `JSON.parse()` should be wrapped in a try/catch.
> 
> More generally, do you have a test page set up with the pixel to test this with?

Doh! And yes we do, thanks to Austin & Kevin.

elmiranda commented:
@woodchuckgym I updated this PR to reflect that the Pixel will do the handling of the stringified JSON, so we don't need to do any try/catching here.

woodchuckgym reviewed (approved):
sgtm.  lmk how testing goes!

woodchuckgym reviewed (commented):
Hey, what happened in all these recent commits?  The image shrunk again, for sure -- which I had asked for changes before on -- but something touched the template file again?

elmiranda commented:
> Hey, what happened in all these recent commits? The image shrunk again, for sure -- which I had asked for changes before on -- but something touched the template file again?

_Copying over from https://github.com/reddit/reddit-gtm-template/pull/22 for posterity:_ 

Hey! I rebased the PR and removed a variable from tests that was added initially, but not removed after some of the later changes. Then anything else was just trying to improve the image size. I reverted the change that Monica hadn't seen.
